File: app/src/main/java/com/worldventures/dreamtrips/modules/reptools/service/SuccessStoriesInteractor.java
package com.worldventures.dreamtrips.modules.reptools.service;

import com.worldventures.dreamtrips.core.janet.SessionActionPipeCreator;
import com.worldventures.dreamtrips.modules.reptools.service.command.GetSuccessStoriesCommand;
import com.worldventures.dreamtrips.modules.reptools.service.command.LikeSuccessStoryCommand;
import com.worldventures.dreamtrips.modules.reptools.service.command.UnlikeSuccessStoryCommand;

import io.techery.janet.ActionPipe;
import rx.schedulers.Schedulers;

public class SuccessStoriesInteractor {

   private ActionPipe<GetSuccessStoriesCommand> getSuccessStoriesPipe;
   private ActionPipe<LikeSuccessStoryCommand> likeSuccessStoryPipe;
   private ActionPipe<UnlikeSuccessStoryCommand> unlikeSuccessStoryPipe;

   public SuccessStoriesInteractor(SessionActionPipeCreator sessionActionPipeCreator) {
      getSuccessStoriesPipe = sessionActionPipeCreator.createPipe(GetSuccessStoriesCommand.class, Schedulers.io());
      likeSuccessStoryPipe = sessionActionPipeCreator.createPipe(LikeSuccessStoryCommand.class, Schedulers.io());
      unlikeSuccessStoryPipe = sessionActionPipeCreator.createPipe(UnlikeSuccessStoryCommand.class, Schedulers.io());
   }

   public ActionPipe<GetSuccessStoriesCommand> getSuccessStoriesPipe() {
      return getSuccessStoriesPipe;
   }

   public ActionPipe<LikeSuccessStoryCommand> likeSuccessStoryPipe() {
      return likeSuccessStoryPipe;
   }

   public ActionPipe<UnlikeSuccessStoryCommand> unlikeSuccessStoryPipe() {
      return unlikeSuccessStoryPipe;
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/service/reptools/SuccessStoriesInteractor.kt
package com.worldventures.dreamtrips.social.service.reptools

import com.worldventures.core.janet.SessionActionPipeCreator
import com.worldventures.dreamtrips.social.service.reptools.command.GetSuccessStoriesCommand
import com.worldventures.dreamtrips.social.service.reptools.command.LikeSuccessStoryCommand
import com.worldventures.dreamtrips.social.service.reptools.command.RefreshSuccessStoriesCommand
import com.worldventures.dreamtrips.social.service.reptools.command.SuccessStoriesCommand
import com.worldventures.dreamtrips.social.service.reptools.command.UnlikeSuccessStoryCommand
import com.worldventures.dreamtrips.social.service.reptools.command.UpdateSuccessStoryLikeStatusCommand
import rx.schedulers.Schedulers

class SuccessStoriesInteractor(pipeCreator: SessionActionPipeCreator) {

   val getSuccessStoriesPipe = pipeCreator.createPipe(GetSuccessStoriesCommand::class.java, Schedulers.io())
   val likeSuccessStoryPipe = pipeCreator.createPipe(LikeSuccessStoryCommand::class.java, Schedulers.io())
   val unlikeSuccessStoryPipe = pipeCreator.createPipe(UnlikeSuccessStoryCommand::class.java, Schedulers.io())
   val successStoriesPipe = pipeCreator.createPipe(SuccessStoriesCommand::class.java, Schedulers.io())
   val updateLikeStatusPipe = pipeCreator.createPipe(UpdateSuccessStoryLikeStatusCommand::class.java, Schedulers.io())

   init {
      likeSuccessStoryPipe.observeSuccess()
            .map { it.result }
            .subscribe { successStoriesPipe.send(UpdateSuccessStoryLikeStatusCommand(it, true)) }

      unlikeSuccessStoryPipe.observeSuccess()
            .map { it.result }
            .subscribe { successStoriesPipe.send(UpdateSuccessStoryLikeStatusCommand(it, false)) }

      getSuccessStoriesPipe.observeSuccess()
            .map { it.result }
            .subscribe { successStoriesPipe.send(RefreshSuccessStoriesCommand(it)) }
   }
}
